### PR TITLE
sink(cdc): clean backends if table sink is stuck too long

### LINKED ISSUE #9542: kafka dml sink can lost callbacks
### What did you do?

1. Create a changefeed with kafka sink.
2. Inject some chaos to the kafka sink, for example, network partition.

Table sink callbacks can be lost. Then table progress can't be advanced any more.

### What did you expect to see?

_No response_

### What did you see instead?

TiCDC table progress is stuck.

### Versions of the cluster

TiCDC 6.5, 7.1 and master branches have the problem.

### PR DESCRIPTION
**Pull request by hicqu**
### What problem does this PR solve?

Issue Number: close #9542  

### What is changed and how it works?

Close DML sink backend after a given duration, if checkpoint can't be advanced correctly as expected.
And then re-establish the DML sink to continue sink to the downstream.

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/processor/sinkmanager/table_sink_wrapper.go [hunk 1]
```diff
@@ -183,25 +191,32 @@ func (t *tableSinkWrapper) updateReceivedSorterResolvedTs(ts model.Ts) {
...
 }
 
-func (t *tableSinkWrapper) getCheckpointTs() model.ResolvedTs {
-	t.tableSinkMu.RLock()
-	defer t.tableSinkMu.RUnlock()
-	if t.tableSink != nil {
-		checkpointTs := t.tableSink.GetCheckpointTs()
-		if t.tableSinkCheckpointTs.Less(checkpointTs) {
-			t.tableSinkCheckpointTs = checkpointTs
+func (t *tableSinkWrapper) getCheckpointTs() (model.ResolvedTs, uint64, time.Time) {
```
Comment (nongfushanquan) [inline cdc/processor/sinkmanager/table_sink_wrapper.go:203]: please add some comments about the output value

File: cdc/processor/sinkmanager/manager.go [hunk 2]
```diff
@@ -86,8 +86,12 @@ type SinkManager struct {
 	sourceManager *sourcemanager.SourceManager
```
Comment (3AceShowHand) [inline cdc/processor/sinkmanager/manager.go:86]: There are 2 constant variable `maxRetryDuration` and `errGCInterval` not used, can you remove them ?

File: cdc/processor/sinkmanager/manager.go [hunk 3]
```diff
@@ -304,46 +302,91 @@ func (m *SinkManager) Run(ctx context.Context, warnings ...chan<- error) (err er
...
+		for range m.sinkFactory.errors {
+		}
+		m.sinkFactory.errors = nil
+	}
+}
+
+func (m *SinkManager) putSinkFactoryError(err error, version uint64) bool {
+	m.sinkFactory.Lock()
+	defer m.sinkFactory.Unlock()
+	skipped := true
```
Comment (3AceShowHand) [inline cdc/processor/sinkmanager/manager.go:373]: how about init `skipped` to `false`, no need to flip the skipped as the return value

Comment (3AceShowHand) [reply cdc/processor/sinkmanager/manager.go:373]: please please add some comment about skipped, what does it means to the caller ?

File: cdc/processor/sinkmanager/manager.go [hunk 4]
```diff
@@ -304,46 +302,91 @@ func (m *SinkManager) Run(ctx context.Context, warnings ...chan<- error) (err er
...
+		default:
+		}
 	}
+	log.Info("Sink manager tries to put an sink error",
+		zap.String("namespace", m.changefeedID.Namespace),
+		zap.String("changefeed", m.changefeedID.ID),
+		zap.Bool("skipped", skipped),
+		zap.String("error", err.Error()))
+
+	return !skipped
```
Comment (3AceShowHand) [inline cdc/processor/sinkmanager/manager.go:387]: it looks this return value is not used by the caller

Comment (hicqu) [reply cdc/processor/sinkmanager/manager.go]: Yes, it's removed now.

File: cdc/processor/sinkmanager/manager.go [hunk 5]
```diff
@@ -86,8 +86,12 @@ type SinkManager struct {
 	sourceManager *sourcemanager.SourceManager
 
 	// sinkFactory used to create table sink.
-	sinkFactory   *factory.SinkFactory
-	sinkFactoryMu sync.Mutex
+	sinkFactory struct {
+		sync.Mutex
+		f       *factory.SinkFactory
+		version uint64
```
Comment (3AceShowHand) [inline cdc/processor/sinkmanager/manager.go:90]: please add a comment about the `version` to indicate why it's needed ?

File: cdc/processor/sinkmanager/manager.go [hunk 6]
```diff
@@ -931,9 +975,24 @@ func (m *SinkManager) GetTableStats(span tablepb.Span) TableStats {
 	}
 	tableSink := value.(*tableSinkWrapper)
 
-	checkpointTs := tableSink.getCheckpointTs()
+	checkpointTs, version, advanced := tableSink.getCheckpointTs()
 	m.sinkMemQuota.Release(span, checkpointTs)
 	m.redoMemQuota.Release(span, checkpointTs)
+
+	stuckCheck := time.Duration(*m.changefeedInfo.Config.Sink.AdvanceTimeout) * time.Second
+	if advanced.After(time.Unix(0, 0)) && time.Since(advanced) > stuckCheck &&
```
Comment (3AceShowHand) [inline cdc/processor/sinkmanager/manager.go:979]: It looks `advanced.After(time.Unix(0, 0))` is always false ?

File: cdc/api/v2/model.go [hunk 7]
```diff
@@ -788,6 +794,7 @@ type SinkConfig struct {
 	KafkaConfig                      *KafkaConfig        `json:"kafka_config,omitempty"`
 	MySQLConfig                      *MySQLConfig        `json:"mysql_config,omitempty"`
 	CloudStorageConfig               *CloudStorageConfig `json:"cloud_storage_config,omitempty"`
+	AdvanceTimeout                   *uint               `json:"advance_timeout,omitempty"`
```
Comment (sdojjy) [inline cdc/api/v2/model.go:797]: maybe it's better to add a time unit to the field,  something like AdvenceTiimeoutInSecond

File: cdc/processor/sinkmanager/manager.go [hunk 8]
```diff
@@ -931,9 +975,24 @@ func (m *SinkManager) GetTableStats(span tablepb.Span) TableStats {
 	}
 	tableSink := value.(*tableSinkWrapper)
 
-	checkpointTs := tableSink.getCheckpointTs()
+	checkpointTs, version, advanced := tableSink.getCheckpointTs()
 	m.sinkMemQuota.Release(span, checkpointTs)
 	m.redoMemQuota.Release(span, checkpointTs)
+
+	stuckCheck := time.Duration(*m.changefeedInfo.Config.Sink.AdvanceTimeout) * time.Second
```
[the remark this answers is not in the document]
Comment (hicqu) [reply cdc/processor/sinkmanager/manager.go:978]: It's a trivial logic, so I prefer to calculate it here.

File: cdc/processor/sinkmanager/manager.go (hunk 6 above)
Comment (hicqu) [reply cdc/processor/sinkmanager/manager.go:979]: `advance` can be equal to `time.Unix(0, 0)`, which means the table doesn't have a valid table sink at the time.

**Review by 3AceShowHand — APPROVED** [review]

**Review by sdojjy — APPROVED** [review]

**Pull request merged**